src/proto/hybrid_vec.rs

```diff
@@ -0,0 +1,1832 @@
+        dst.set_len(elts);
+        ptr::copy_nonoverlapping_memory(dst.as_mut_ptr(), ptr, elts);
+        dst
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        match *self {
+            HybridVec::Heap {  ptr: _, len, cap: _ } => len,
+            HybridVec::Stack { array: _, len } => len,    
+        }    
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        match *self {
+            HybridVec::Heap { ptr: _, len: _, cap } => cap,
+            HybridVec::Stack { array: _, len: _ } => ARRAY_SIZE,    
+        }    
+    }
+    
+    pub fn insert(&mut self, index: usize, element: T) {
+        let len = self.len();
+        assert!(index <= len);
+        // space for the new element
+        self.reserve(1);
+
+        unsafe { // infallible
+            // The spot to put the new value
+            {
+                let p = self.as_mut_ptr().offset(index as isize);
+                // Shift everything over to make space. (Duplicating the
+                // `index`th element into two consecutive places.)
+                ptr::copy_memory(p.offset(1), &*p, len - index);
+                // Write it in, overwriting the first copy of the `index`th
+                // element.
+                ptr::write(&mut *p, element);
+            }
+            self.set_len(len + 1);
+        }
+    }
+
+    pub fn remove(&mut self, index: usize) -> T {
+        let len = self.len();
+        assert!(index < len);
+        unsafe { // infallible
+            let ret;
+            {
+                // the place we are taking from.
+                let ptr = self.as_mut_ptr().offset(index as isize);
+                // copy it out, unsafely having a copy of the value on
+                // the stack and in the vector at the same time.
+                ret = ptr::read(ptr as *const T);
+
+                // Shift everything down to fill in that spot.
+                ptr::copy_memory(ptr, &*ptr.offset(1), len - index - 1);
+            }
+            self.set_len(len - 1);
+            ret
+        }
+    }
+    
+    //#[inline]
+    pub fn push(&mut self, value: T) {
+        if mem::size_of::<T>() == 0 {
+            // zero-size types consume no memory, so we can't rely on the
+            // address space running out.
+            // This will automatically use the heap variant.
+            unsafe {                
+                let new_len = self.len().checked_add(1).expect("length overflow");
+                self.set_len(new_len);
+                mem::forget(value); 
+            }
+            return
+        }
+        
+        // `HybridVec::push()` had some weird reallocation logic that wasn't explained in the code
+        // and didn't seem to change much except the growth rate of `HybridVec`.
+        // This should suffice unless the old behavior is shown to be preferiential.
+        self.reserve(1);
```

Comment:
This should probably choose `Vec::push()`'s behavior for heap-resizing and then use this for stack-resizing (effectively a no-op if the new capacity is within the bounds of the on-stack array).